ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/AddFunctionToSignatureAction.fs

```diff
@@ -0,0 +1,91 @@
+namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Intentions
+
+open JetBrains.ReSharper.Feature.Services.ContextActions
+open JetBrains.ReSharper.Plugins.FSharp.Psi
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.Tree
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open JetBrains.ReSharper.Psi.ExtensionsAPI
+open JetBrains.ReSharper.Psi.ExtensionsAPI.Tree
+open JetBrains.ReSharper.Psi.Tree
+open JetBrains.ReSharper.Resources.Shell
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Services.Util
+
+[<ContextAction(Name = "AddFunctionToSignatureFile", Group = "F#", Description = "Add function to signature file")>]
+type AddFunctionToSignatureFileAction(dataProvider: FSharpContextActionDataProvider) =
+    inherit ContextActionBase()
+    
+    let getSignatureLocation () =
+        let letBindings = dataProvider.GetSelectedElement<ILetBindings>()
+        if isNull letBindings then None else
+
+        let bindings = letBindings.Bindings
+        if bindings.Count <> 1 then None else
+        
+        match bindings.First().HeadPattern.As<IReferencePat>() with
+        | null -> None
+        | pat -> pat.GetFcsSymbol().SignatureLocation
+
+    let addChildAfter (newlineNode: ITreeNode) (insertNewline: bool) (child: ITreeNode) (anchor: ITreeNode) =
+        let anchor =
+            let anchor =
+                if anchor.NodeType <> newlineNode.NodeType then
+                    ModificationUtil.AddChildAfter(anchor, newlineNode)
+                else
+                    anchor
+
+            if insertNewline then
+                ModificationUtil.AddChildAfter(anchor, newlineNode)
+            else
+                anchor
+
+        ModificationUtil.AddChildAfter(ModificationUtil.AddChildAfter(anchor, child), newlineNode)
+    
+    override x.Text = "Add to signature file"
+    override this.IsAvailable _ =
+        let currentFSharpFile = dataProvider.PsiFile
+        let fcsService = currentFSharpFile.FcsCheckerService
+        let hasSignature = fcsService.FcsProjectProvider.HasPairFile dataProvider.SourceFile
+        if not hasSignature then false else
+        
+        match getSignatureLocation () with
+        | None -> false
+        | Some range -> not (range.FileName.EndsWith(".fsi"))
+
+    override this.ExecutePsiTransaction(solution, progress) =
+        use writeCookie = WriteLockCookie.Create(true)
+        use disableFormatter = new DisableCodeFormatter()
+        let letBindings = dataProvider.GetSelectedElement<ILetBindingsDeclaration>()
+        let moduleDecl = ModuleDeclarationNavigator.GetByMember(letBindings)
+
+        match SignatureFile.tryMkBindingSignature letBindings moduleDecl with
+        | None -> null
+        | Some response ->
+            let sigFile = response.SigFile
```

Comment:
Hey @auduchinok, any advice on handling blank lines?
I'm making a mess in the code here.

I'm faced with scenarios where you want to introduce a blank line (I think).
Like from

```fsharp
module Foo
```

to

```fsharp
module Foo

val bar: int
```

But you don't want this all the time:

Example from
```fsharp
module Foo

open System
```

to

```fsharp
module Foo

open System
open System.Collections.Generic

val bar : IDictionary<string,DateTime>
```

You don't want to add a blank line between the open statements.

Reply:
There's some code in `OpensUtil.addOpen` and places like `GenerateOverrides.addEmptyLineAfterIfNeeded`. We could copy or generalize a few pieces from them.

The best approach would be to finally finish the formatter for top level members, and it'd do these things automatically when adding a member node. We could try to enable it here just to see where/if it breaks. 🙂